wannierberri/__kubo.py

```diff
@@ -149,26 +177,59 @@ def opt_conductivity(data, omega=0, mu=0, kBT=0, smr_fixed_width=0.1, smr_type='
-
-    # return result dictionary
-    return result.EnergyResultDict({
-        'sym':  result.EnergyResult(omega, sigma_sym, TRodd=False, Iodd=False, rank=2),
-        'asym': result.EnergyResult(omega, sigma_asym, TRodd=True, Iodd=False, rank=2)
-    }) # the proper smoother is set later for both elements
+        if conductivity_type == 'kubo':
+            # real part of energy fraction [iw, n, m]
+            re_efrac = delta_arg/(delta_arg**2 + eta**2)
+
+            # temporary variables for computing conductivity tensor
+            tmp1 = dfE * dE
+            tmp2 = np.einsum('nma,mnb->nmab', A, A)
+            tmp3 = tmp1[:, :, np.newaxis, np.newaxis] * tmp2
+
+            # Hermitian part of the conductivity tensor
+            sigma_H += -1 * pi * pre_fac * kubo_sum_elements(tmp3, delta, data.num_wann)
+            # anti-Hermitian part of the conductivity tensor
+            sigma_AH += 1j * pre_fac * kubo_sum_elements(tmp3, re_efrac, data.num_wann)
+
+        elif conductivity_type == 'SHC':
+            delta_minus = np.copy(delta)
+            delta_arg = dE[np.newaxis,:,:] + omega[:,np.newaxis,np.newaxis]
+            if smr_type == 'Lorentzian':
+                delta = Lorentzian(delta_arg, eta)
+            elif smr_type == 'Gaussian':
+                delta = Gaussian(delta_arg, eta, adpt_smr)
+            else:
+                cprint("Invalid smearing type. Fallback to Lorentzian", 'red')
+                delta = Lorentzian(delta_arg, eta)
+            delta_plus = np.copy(delta)
+            cfac2 = delta_minus - delta_plus
+            cfac1 = np.real(-dE[np.newaxis,:,:]/(dE[np.newaxis,:,:]**2-(omega[:,np.newaxis,np.newaxis]+1j*eta)**2))
+            temp1 = dfE[np.newaxis,:,:]*cfac1
+            temp2 = dfE[np.newaxis,:,:]*cfac2
+            imAB = np.imag(np.einsum('nmac,mnb->nmabc',A,B))
+            sigma_H += 1j * pi * pre_fac * kubo_sum_elements(imAB, temp2, data.num_wann) / 4.0
+            sigma_AH += pre_fac * kubo_sum_elements(imAB, temp1, data.num_wann) / 2.0
+
+    if conductivity_type == 'kubo':
+        # TODO: optimize by just storing independent components or leave it like that?
+        # 3x3 tensors [iw, a, b]
+        sigma_sym = np.real(sigma_H) + 1j * np.imag(sigma_AH) # symmetric (TR-even, I-even)
+        sigma_asym = np.real(sigma_AH) + 1j * np.imag(sigma_H) # ansymmetric (TR-odd, I-even)
+
+        # return result dictionary
+        return result.EnergyResultDict({
+            'sym':  result.EnergyResult(omega, sigma_sym, TRodd=False, Iodd=False, rank=rank),
+            'asym': result.EnergyResult(omega, sigma_asym, TRodd=True, Iodd=False, rank=rank)
+        }) # the proper smoother is set later for both elements
+
+    elif conductivity_type == 'SHC':
+        sigma_SHC = np.real(sigma_AH) + 1j * np.imag(sigma_H)
+        return result.EnergyResultDict({
+            'freqscan': result.EnergyResult(omega, sigma_SHC, TRodd=False, Iodd=False, rank=rank)
```

Comment:
Why I used EnergyResultDict is to add 'fermiscan' function later.

Reply:
You mean - to scan Fermi energy with a constant frequency? 
I would say it is better to design a way to scan both of them simultaneously. That's smth thatshould be done eventually.  I guess it will requre a new class, smth like 'EfOmegaResult'